kotsadm/pkg/socketservice/socketservice.go

```diff
@@ -0,0 +1,543 @@
+	ClearNamespaces      []string `json:"clear_namespaces"`
+	ClearPVCs            bool     `json:"clear_pvcs"`
+	AnnotateSlug         bool     `json:"annotate_slug"`
+}
+
+type AppInformersArgs struct {
+	AppID     string   `json:"app_id"`
+	Informers []string `json:"informers"`
+}
+
+type SupportBundleArgs struct {
+	URI string `json:"uri"`
+}
+
+var server *socket.Server
+var clusterSocketHistory = []ClusterSocket{}
+var socketMtx sync.Mutex
+
+func Start() *socket.Server {
+	server = socket.NewServer(transport.GetDefaultWebsocketTransport())
+
+	server.On(socket.OnConnection, func(c *socket.Channel, args interface{}) {
+		socketMtx.Lock()
+		defer socketMtx.Unlock()
+
+		d, err := downstream.GetFromDeployToken(c.RequestURL().Query().Get("token"))
+		if err != nil {
+			logger.Error(errors.Wrap(err, "failed to get cluster id from deploy token"))
+			return
+		}
+
+		logger.Info(fmt.Sprintf("cluster %s connected", d.ClusterID))
+		c.Join(d.ClusterID)
+
+		clusterSocket := ClusterSocket{
+			ClusterID:             d.ClusterID,
+			SocketID:              c.Id(),
+			SentPreflightURLs:     make(map[string]bool, 0),
+			LastDeployedSequences: make(map[string]int64, 0),
+		}
+		clusterSocketHistory = append(clusterSocketHistory, clusterSocket)
+	})
+
+	server.On(socket.OnDisconnection, func(c *socket.Channel) {
+		socketMtx.Lock()
+		defer socketMtx.Unlock()
+
+		updatedClusterSocketHistory := []ClusterSocket{}
+		for _, clusterSocket := range clusterSocketHistory {
+			if clusterSocket.SocketID != c.Id() {
+				updatedClusterSocketHistory = append(updatedClusterSocketHistory, clusterSocket)
+			}
+		}
+		clusterSocketHistory = updatedClusterSocketHistory
+	})
+
+	startLoop(deployLoop, 1)
+	startLoop(supportBundleLoop, 1)
+	startLoop(restoreLoop, 1)
+
+	return server
+}
+
+func startLoop(fn func(), intervalInSeconds time.Duration) {
+	go func() {
+		for {
+			fn()
+			time.Sleep(time.Second * intervalInSeconds)
+		}
+	}()
+}
+
+func deployLoop() {
+	for _, clusterSocket := range clusterSocketHistory {
+		apps, err := app.ListInstalledForDownstream(clusterSocket.ClusterID)
+		if err != nil {
+			logger.Error(errors.Wrap(err, "failed to list installed apps for downstream"))
+			continue
+		}
+
+		for _, a := range apps {
+			if a.RestoreInProgressName != "" {
+				continue
+			}
+
+			deployedVersion, err := downstream.GetCurrentVersion(a.ID, clusterSocket.ClusterID)
+			if err != nil {
+				logger.Error(errors.Wrap(err, "failed to get current downstream version"))
+				continue
+			}
+
+			if deployedVersion == nil {
+				continue
+			}
+
+			if value, ok := clusterSocket.LastDeployedSequences[a.ID]; ok && value == deployedVersion.ParentSequence {
+				// this version is already the currently deployed version
+				continue
+			}
+
+			d, err := downstream.Get(clusterSocket.ClusterID)
+			if err != nil {
+				logger.Error(errors.Wrap(err, "failed to get downstream"))
+				continue
+			}
+
+			var deployError error
+			defer func() {
```

Comment:
this defer function will not get called until the deploy loop ends. does it ever? also perhaps this warrants deploy() being its own function?

Reply:
good catch

Reply:
the deployLoop function does end, but it gets called every 1 second. but i think the location of this should be changed, not sure where to yet

Reply:
moving the deploy to it's own function is a good suggestion that would solve this 👍